inference-engine/src/multi_device/multi_device_exec_network.cpp

```diff
@@ -60,50 +90,150 @@ MultiDeviceExecutableNetwork::MultiDeviceExecutableNetwork(const DeviceMap<Infer
+                {
+                    auto capturedTask = std::move(workerRequestPtr->_task);
+                    capturedTask();
+                }
+                // try to return the request to the idle list (fails if the overall object destruction has began)
+                if (idleGuard.Release()->try_push(workerRequestPtr)) {
+                    // let's try to pop a task, as we know there is at least one idle request, schedule if succeeded
+                    // if no device-agnostic tasks, let's try pop the device specific task, schedule if succeeded
+                    Task t;
+                    if (_inferPipelineTasks.try_pop(t))
+                        ScheduleToWorkerInferRequest(std::move(t));
+                    else if (_inferPipelineTasksDeviceSpecific[device]->try_pop(t))
+                        ScheduleToWorkerInferRequest(std::move(t), device);
+                }
+            });
+    }
+}
+
+MultiDeviceExecutableNetwork::MultiDeviceExecutableNetwork(const std::string&                         modelPath,
+                                                           const InferenceEngine::CNNNetwork&         network,
+                                                           const std::map<std::string, std::string>&  config,
+                                                           MultiDeviceInferencePlugin*                plugin)
+                                                           : _multiPlugin(plugin)
+                                                           , _workModeIsAUTO(true) {
+    if (_multiPlugin->GetCore() == nullptr) {
+        IE_THROW() << "Please, work with MULTI device via InferencEngine::Core object";
+    }
+
+    if (modelPath.empty() && network.getFunction() == nullptr) {
+        IE_THROW() << "MULTI device supports just ngraph network representation";
+    }
+
+    auto strDevices = _multiPlugin->GetDeviceList(config);
+    // collect the settings that are applicable to the devices we are loading the network to
+    _config[MultiDeviceConfigParams::KEY_MULTI_DEVICE_PRIORITIES] = strDevices;
+
+    auto metaDevices = _multiPlugin->ParseMetaDevices(strDevices, config);
+    _devicePrioritiesInitial = metaDevices;
+    _devicePriorities = metaDevices;
+
+    // init worker queue
+    for (auto&& device : metaDevices) {
+        _idleWorkerRequests[device.deviceName];
+    }
+
+    auto core = _multiPlugin->GetCore(); // shared_ptr that holds the Core while the lambda below (which captures that by val) works
+    auto LoadNetworkAsync =
+        [this, core, modelPath, network](const std::string& device) -> SoExecutableNetworkInternal {
+            SoExecutableNetworkInternal executableNetwork;
+            if (!modelPath.empty()) {
+                executableNetwork = core->LoadNetwork(modelPath, device, {});
+            } else {
+                executableNetwork = core->LoadNetwork(network, device, {});
+            }
+
+            GenerateWorkers(device, executableNetwork);
```

Comment:
as GenerateWorkers can be called in parallel pls make sure to protect it with a mutex, or at least to protect certain places in that with the mutexes like inserting into the _workerRequests via the [], insertting to  the _inferPipelineTasksDeviceSpecific etc

Reply:
Thanks Maxim, you're right, that is a bug, I will fix that

Reply:
these containers is initialized at the begin of thread task now, fixed this potential risk